### PR TITLE
[CLI] Verify network with provider

### LINKED ISSUE #2113: Default arguments from scarb.toml
As discussed with @war-in , let's simplify ui/ux further, and try to fetch default values for things like rpc_url from scarb.toml.
That would imply, that we no longer need to verify network with provider (as per todo 2113 in code), as the proposed template would be:

```
[tool.protostar.cli.mainnet]
...

[tool.protostar.cli.devnet]
...
```

Possibly to do after mvp for scarb integration

### PR DESCRIPTION
**Pull request by THenry14**
### Checklist
- [X] Relevant issue is linked
- [X] Docs updated/issue for docs created
- [X] Added relevant tests

Closes https://github.com/software-mansion/protostar/issues/2113

### CODE REVIEW

File: cast/src/lib.rs [hunk 1]
```diff
@@ -66,9 +67,17 @@ impl Network {
...
-pub fn get_provider(url: &str) -> Result<JsonRpcClient<HttpTransport>> {
+pub fn get_provider(url: &str, network: &Network) -> Result<JsonRpcClient<HttpTransport>> {
     let parsed_url = Url::parse(url)?;
     let provider = JsonRpcClient::new(HttpTransport::new(parsed_url));
+
+    let rt = Runtime::new()?;
+    let provider_chain_id = rt.block_on(provider.chain_id())?;
+    let cli_chain_id = network.get_chain_id();
+    if provider_chain_id != cli_chain_id {
+        bail!("Networks mismatch: requested network is different than provider network!")
```
Comment (war-in) [inline cast/src/lib.rs:78]: Should we migrate all of `Err(anyhow!(...))` to `bail!`?

File: cast/src/lib.rs [hunk 2]
```diff
@@ -66,9 +67,17 @@ impl Network {
...
 }
 
-pub fn get_provider(url: &str) -> Result<JsonRpcClient<HttpTransport>> {
+pub fn get_provider(url: &str, network: &Network) -> Result<JsonRpcClient<HttpTransport>> {
     let parsed_url = Url::parse(url)?;
     let provider = JsonRpcClient::new(HttpTransport::new(parsed_url));
+
+    let rt = Runtime::new()?;
+    let provider_chain_id = rt.block_on(provider.chain_id())?;
+    let cli_chain_id = network.get_chain_id();
```
Comment (war-in) [inline cast/src/lib.rs:76]: I've started wondering if we could get rid of the `network` parameter entirely.
Since we are calling network for its chain_id here we do not need `network` parameter anymore.
It's useful only when getting an account but we could change the logic of accessing the network's value.
What do you think?

File: cast/tests/e2e/main.rs [hunk 3]
```diff
@@ -1 +1,2 @@
+mod helper_functions;
 pub(crate) mod helpers;
```
Comment (war-in) [inline cast/tests/e2e/main.rs:2]: I think we should create a command to invoke all tests at once

File: cast/src/lib.rs (hunk 2 above)
Comment (war-in) [reply cast/src/lib.rs:76]: I think UX would be improved because it's one parameter less :)

File: cast/src/lib.rs (hunk 1 above)
Comment (THenry14) [reply cast/src/lib.rs:78]: let's. do you want me to handle this, or do you want to do this?

File: cast/src/lib.rs (hunk 2 above)
Comment (THenry14) [reply cast/src/lib.rs:76]: hmm, makes sense. I guess we should still get network from scarb.toml though, right? in order to compare the networks and throw an error when there is mismatch, to avoid accidental declare/deploy to wrong network?

File: cast/tests/e2e/main.rs (hunk 3 above)
Comment (THenry14) [reply cast/tests/e2e/main.rs:2]: can you clarify this, please? I'm not sure I follow

Comment (war-in) [reply cast/tests/e2e/main.rs:2]: My bad. I forgot there is `cargo test` 🤦

File: cast/src/lib.rs (hunk 1 above)
Comment (war-in) [reply cast/src/lib.rs:78]: I will in a separate PR after #2146 is merged

File: cast/tests/e2e/helper_functions.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
+use crate::helpers::devnet::URL;
```
Comment (karol-bisztyga) [inline cast/tests/e2e/helper_functions.rs:1]: Sorry but I do not understand this file, it is very misleading. You called it `helper_functions.rs` but all I see here are tests. It doesn't make sense
Could you explain why did you move all these tests to a separate file in the first place? It is basically against the rust convention that says we should have such tests in the same file as functions. Can we move them back there?

File: cast/tests/data/accounts/user1_representation [hunk 5]
```diff
@@ -1 +1 @@
...
\ No newline at end of file
+SingleOwnerAccount { provider: JsonRpcClient { transport: HttpTransport { client: Client { accepts: Accepts, proxies: [Proxy(System({}), None)], referer: true, default_headers: {"accept": "*/*"} }, url: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(5055), path: "/rpc", query: None, fragment: None } } }, signer: LocalWallet { private_key: SigningKey { secret_scalar: FieldElement { inner: 0x0000000000000000000000000000000055ae34c86281fbd19292c7e3bfdfceb4 } } }, address: FieldElement { inner: 0x0112b61669a84940af30544786a6b615755b192173c4665a7d38fb3aa0541cac }, chain_id: FieldElement { inner: 0x0000000000000000000000000000000000000000000000534e5f474f45524c49 }, block_id: Tag(Latest) }
```
Comment (karol-bisztyga) [inline cast/tests/data/accounts/user1_representation:1]: can you explain why this changed?

File: cast/tests/e2e/helper_functions.rs (hunk 4 above)
Comment (karol-bisztyga) [reply cast/tests/e2e/helper_functions.rs:1]: Also, tests here don't look like e2e tests, are you sure they are e2e?

**Review by karol-bisztyga — CHANGES_REQUESTED** [review]

File: cast/tests/data/accounts/user1_representation (hunk 5 above)
Comment (THenry14) [reply cast/tests/data/accounts/user1_representation:1]: I have moved those tests to e2e as they are end to end. becasue of that devnet is no longer mocked, and thus we are using the port from config (which is 5055)

File: cast/tests/e2e/helper_functions.rs (hunk 4 above)
Comment (THenry14) [reply cast/tests/e2e/helper_functions.rs:1]: Sure, happy to help.

I have named this file helper_functions, as it is supposed to test all the helper functions we have (and I was not sure I can name it lib to idjicate those helper functions are from lib exclusively). This resides in test directory, where all the test files are present. If you have a better idea how to name it, let me know :) 

>Could you explain why did you move all these tests to a separate file in the first place? It is basically against the rust convention that says we should have such tests in the same file as functions. Can we move them back there?

the convention is that all the unit tests for specific file are to be in this file, but e2e tests should not be there + I nned them to be in e2e directory since 1. they are e2e 2. They require devnet, which is assured only for tests in e2e directory. We cannot, and we should not move them back, becasue they will break.

> 
Also, tests here don't look like e2e tests, are you sure they are e2e?

they are, they all setup up a provider, which makes them e2e.

**Comment by THenry14** [comment]
Convertged to draft, because this PR will change a lot https://github.com/software-mansion/protostar/issues/2113

File: cast/tests/e2e/helper_functions.rs (hunk 4 above)
Comment (karol-bisztyga) [reply cast/tests/e2e/helper_functions.rs:1]: Discussed offline, and agreed we should move this to the `integration` folder as they are indeed integration tests, not e2e.

> If you have a better idea how to name it, let me know :)

Sure, I think this should be named `cast/tests/integration/lib_tests.rs` because it tests functions from `lib.rs`. `helper_functions.rs` doesn't tell much tbh.

Comment (karol-bisztyga) [reply cast/tests/e2e/helper_functions.rs:1]: > the convention is that all the unit tests for specific file are to be in this file, but e2e tests should not be there

Yes, this is correct, thanks for clarifying, that also applies to integration tests.
